File: src/services/taskHistoryService.ts
import type {
  ImportResult,
} from '../types/import'
import type {
  LanguageCode,
  MediaType,
  ProjectFile,
  ProjectState,
  ProjectStatus,
  SubtitleSegment,
} from '../types/models'
import type {
  TaskEngineType,
  TaskHistoryRecord,
  TaskHistoryResponse,
  TaskLogEntry,
  TaskLogLevel,
  TaskMode,
  TaskStatus,
} from '../types/tasks'
import { requestJson } from './backendClient'

const VALID_TASK_MODES = new Set<TaskMode>(['extractAndTranslate', 'translateSubtitle'])
const VALID_TASK_ENGINE_TYPES = new Set<TaskEngineType>([
  'cloudTranscription',
  'localTranscription',
  'subtitleImport',
])
const VALID_TASK_STATUSES = new Set<TaskStatus>([
  'queued',
  'transcribing',
  'translating',
  'editing',
  'exporting',
  'done',
  'error',
  'cancelled',
])
const VALID_TASK_LOG_LEVELS = new Set<TaskLogLevel>(['info', 'warning', 'error'])
const VALID_PROJECT_STATUSES = new Set<ProjectStatus>([
  'idle',
  'transcribing',
  'translating',
  'exporting',
  'done',
  'error',
])
const VALID_MEDIA_TYPES = new Set<MediaType>(['video', 'audio', 'subtitle'])
const VALID_LANGUAGE_CODES = new Set<LanguageCode>(['auto', 'zh-CN', 'en', 'ja', 'ko'])

function readString(value: unknown, fallback = ''): string {
  return typeof value === 'string' ? value : fallback
}

function readNullableString(value: unknown): string | null {
  return typeof value === 'string' && value.trim().length > 0 ? value : null
}

function readStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return []
  }

  return value.filter((item): item is string => typeof item === 'string')
}

function normalizeTaskLogEntry(value: unknown): TaskLogEntry | null {
  if (!value || typeof value !== 'object') {
    return null
  }

  const rawEntry = value as Partial<TaskLogEntry>
  const level = VALID_TASK_LOG_LEVELS.has(rawEntry.level as TaskLogLevel)
    ? (rawEntry.level as TaskLogLevel)
    : 'info'

  return {
    logId: readString(rawEntry.logId, `log-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`),
    timestamp: readString(rawEntry.timestamp, new Date().toISOString()),
    level,
    message: readString(rawEntry.message, 'LinguaSub restored an older task log entry.'),
    details: readNullableString(rawEntry.details),
  }
}

function normalizeLanguageCode(value: unknown, fallback: LanguageCode): LanguageCode {
  return VALID_LANGUAGE_CODES.has(value as LanguageCode)
    ? (value as LanguageCode)
    : fallback
}

function normalizeSubtitleSegment(value: unknown): SubtitleSegment | null {
  if (!value || typeof value !== 'object') {
    return null
  }

  const rawSegment = value as Partial<SubtitleSegment>
  const id = readString(rawSegment.id)
  if (!id) {
    return null
  }

  return {
    id,
    start: typeof rawSegment.start === 'number' ? rawSegment.start : 0,
    end:
      typeof rawSegment.end === 'number'
        ? rawSegment.end
        : typeof rawSegment.start === 'number'
          ? rawSegment.start
          : 0,
    sourceText: readString(rawSegment.sourceText),
    translatedText: readString(rawSegment.translatedText),
    sourceLanguage: normalizeLanguageCode(rawSegment.sourceLanguage, 'auto'),
    targetLanguage: normalizeLanguageCode(rawSegment.targetLanguage, 'zh-CN'),
  }
}

function normalizeProjectFile(value: unknown): ProjectFile | null {
  if (!value || typeof value !== 'object') {
    return null
  }

  const rawFile = value as Partial<ProjectFile>
  const path = readString(rawFile.path)
  const name = readString(rawFile.name)
  if (!path && !name) {
    return null
  }

  return {
    path,
    name: name || path.split(/[\\/]/).pop() || 'Untitled file',
    mediaType: VALID_MEDIA_TYPES.has(rawFile.mediaType as MediaType)
      ? (rawFile.mediaType as MediaType)
      : 'subtitle',
    extension: readString(rawFile.extension),
    requiresAsr:
      typeof rawFile.requiresAsr === 'boolean'
        ? rawFile.requiresAsr
        : rawFile.mediaType === 'video' || rawFile.mediaType === 'audio',
  }
}

function normalizeProjectState(value: unknown): ProjectState | null {
  if (!value || typeof value !== 'object') {
    return null
  }

  const rawProjectState = value as Partial<ProjectState>
  return {
    currentFile: normalizeProjectFile(rawProjectState.currentFile),
    segments: Array.isArray(rawProjectState.segments)
      ? rawProjectState.segments
          .map((segment) => normalizeSubtitleSegment(segment))
          .filter((segment): segment is SubtitleSegment => segment !== null)
      : [],
    status: VALID_PROJECT_STATUSES.has(rawProjectState.status as ProjectStatus)
      ? (rawProjectState.status as ProjectStatus)
      : 'idle',
    error: readNullableString(rawProjectState.error),
  }
}

function normalizeImportResult(value: unknown): ImportResult | null {
  if (!value || typeof value !== 'object') {
    return null
  }

  const rawImport = value as Partial<ImportResult>
  const currentFile = normalizeProjectFile(rawImport.currentFile)
  if (!currentFile) {
    return null
  }

  return {
    currentFile,
    projectState:
      normalizeProjectState(rawImport.projectState) ?? {
        currentFile,
        segments: [],
        status: 'idle',
        error: null,
      },
    workflow: readStringArray(rawImport.workflow),
    route:
      rawImport.route === 'recognition' || rawImport.route === 'translation'
        ? rawImport.route
        : 'translation',
    shouldSkipTranscription:
      typeof rawImport.shouldSkipTranscription === 'boolean'
        ? rawImport.shouldSkipTranscription
        : currentFile.mediaType === 'subtitle',
    recognitionInput:
      rawImport.recognitionInput &&
      typeof rawImport.recognitionInput === 'object' &&
      typeof rawImport.recognitionInput.mediaPath === 'string'
        ? {
            mediaPath: rawImport.recognitionInput.mediaPath,
            mediaType:
              rawImport.recognitionInput.mediaType === 'video' ||
              rawImport.recognitionInput.mediaType === 'audio'
                ? rawImport.recognitionInput.mediaType
                : 'audio',
            sourceLanguage: normalizeLanguageCode(
              rawImport.recognitionInput.sourceLanguage,
              'auto',
            ),
          }
        : null,
    subtitleInput:
      rawImport.subtitleInput &&
      typeof rawImport.subtitleInput === 'object' &&
      typeof rawImport.subtitleInput.subtitlePath === 'string'
        ? {
            subtitlePath: rawImport.subtitleInput.subtitlePath,
            parser: rawImport.subtitleInput.parser === 'srt' ? 'srt' : 'srt',
            encoding: readString(rawImport.subtitleInput.encoding, 'utf-8'),
          }
        : null,
  }
}

function normalizeTaskHistoryRecord(value: unknown): TaskHistoryRecord | null {
  if (!value || typeof value !== 'object') {
    return null
  }

  const rawTask = value as Partial<TaskHistoryRecord>
  const taskId = readString(rawTask.taskId)
  if (!taskId) {
    return null
  }

  const createdAt = readString(rawTask.createdAt, new Date().toISOString())
  const updatedAt = readString(rawTask.updatedAt, createdAt)
  const taskMode = VALID_TASK_MODES.has(rawTask.taskMode as TaskMode)
    ? (rawTask.taskMode as TaskMode)
    : 'extractAndTranslate'
  const engineType = VALID_TASK_ENGINE_TYPES.has(rawTask.engineType as TaskEngineType)
    ? (rawTask.engineType as TaskEngineType)
    : 'subtitleImport'
  const status = VALID_TASK_STATUSES.has(rawTask.status as TaskStatus)
    ? (rawTask.status as TaskStatus)
    : 'queued'

  return {
    taskId,
    sourceFilePath: readString(rawTask.sourceFilePath),
    sourceFileName: readString(rawTask.sourceFileName, 'Untitled task'),
    taskMode,
    sourceLanguage: readString(rawTask.sourceLanguage, 'auto'),
    targetLanguage: readString(rawTask.targetLanguage, 'zh'),
    outputFormats: readStringArray(rawTask.outputFormats),
    engineType,
    status,
    createdAt,
    updatedAt,
    exportPaths: readStringArray(rawTask.exportPaths),
    errorMessage: readNullableString(rawTask.errorMessage),
    subtitleSummary:
      rawTask.subtitleSummary &&
      typeof rawTask.subtitleSummary === 'object' &&
      typeof rawTask.subtitleSummary.segmentCount === 'number' &&
      typeof rawTask.subtitleSummary.translatedCount === 'number'
        ? {
            segmentCount: rawTask.subtitleSummary.segmentCount,
            translatedCount: rawTask.subtitleSummary.translatedCount,
          }
        : null,
    importSnapshot: normalizeImportResult(rawTask.importSnapshot),
    projectSnapshot: normalizeProjectState(rawTask.projectSnapshot),
    logs: Array.isArray(rawTask.logs)
      ? rawTask.logs
          .map((entry) => normalizeTaskLogEntry(entry))
          .filter((entry): entry is TaskLogEntry => entry !== null)
      : [],
    transcriptionProvider: readNullableString(rawTask.transcriptionProvider),
    transcriptionModelSize: readNullableString(rawTask.transcriptionModelSize),
    transcriptionQualityPreset: readNullableString(rawTask.transcriptionQualityPreset),
    translationProvider: readNullableString(rawTask.translationProvider),
    translationModel: readNullableString(rawTask.translationModel),
    outputMode: readNullableString(rawTask.outputMode),
  }
}

export async function loadTaskHistory(): Promise<TaskHistoryRecord[]> {
  const result = await requestJson<TaskHistoryResponse>('/tasks', {
    method: 'GET',
  })

  if (!result || !Array.isArray(result.tasks)) {
    console.error('LinguaSub task history payload is invalid.', result)
    return []
  }

  return result.tasks
    .map((task) => normalizeTaskHistoryRecord(task))
    .filter((task): task is TaskHistoryRecord => task !== null)
}

export async function upsertTaskHistoryRecord(
  task: TaskHistoryRecord,
): Promise<TaskHistoryRecord> {
  const result = await requestJson<{ task: TaskHistoryRecord }>('/tasks/upsert', {
    method: 'POST',
    body: JSON.stringify({ task }),
  })

  return normalizeTaskHistoryRecord(result.task) ?? task
}

File: backend/app/task_history_service.py
"""Persistent recent-task and diagnostic log storage for LinguaSub."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from time import strftime
from typing import Any, Literal

from .config_service import get_config_path
from .models import JsonModel

TaskMode = Literal["extractAndTranslate", "translateSubtitle"]
TaskEngineType = Literal[
    "cloudTranscription",
    "localTranscription",
    "subtitleImport",
]
TaskStatus = Literal[
    "queued",
    "transcribing",
    "translating",
    "editing",
    "exporting",
    "done",
    "error",
    "cancelled",
]
TaskLogLevel = Literal["info", "warning", "error"]

MAX_HISTORY_ITEMS = 30


class TaskHistoryError(RuntimeError):
    """Base error for recent-task persistence."""


@dataclass(slots=True)
class SubtitleSummary(JsonModel):
    segmentCount: int = 0
    translatedCount: int = 0

    @classmethod
    def from_dict(cls, data: dict[str, Any] | None) -> "SubtitleSummary | None":
        if not isinstance(data, dict):
            return None

        return cls(
            segmentCount=int(data.get("segmentCount", 0) or 0),
            translatedCount=int(data.get("translatedCount", 0) or 0),
        )


@dataclass(slots=True)
class TaskLogEntry(JsonModel):
    logId: str
    timestamp: str
    level: TaskLogLevel
    message: str
    details: str | None = None

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "TaskLogEntry":
        level = str(data.get("level", "info")).strip().lower()
        if level not in {"info", "warning", "error"}:
            level = "info"

        details = data.get("details")
        return cls(
            logId=str(data.get("logId", "")).strip() or f"log-{strftime('%Y%m%d-%H%M%S')}",
            timestamp=str(data.get("timestamp", "")).strip() or "",
            level=level,  # type: ignore[arg-type]
            message=str(data.get("message", "")).strip() or "Task event",
            details=str(details).strip() if isinstance(details, str) and details.strip() else None,
        )


@dataclass(slots=True)
class TaskHistoryRecord(JsonModel):
    taskId: str
    sourceFilePath: str
    sourceFileName: str
    taskMode: TaskMode
    sourceLanguage: str
    targetLanguage: str
    outputFormats: list[str] = field(default_factory=list)
    engineType: TaskEngineType = "subtitleImport"
    status: TaskStatus = "queued"
    createdAt: str = ""
    updatedAt: str = ""
    exportPaths: list[str] = field(default_factory=list)
    errorMessage: str | None = None
    subtitleSummary: SubtitleSummary | None = None
    importSnapshot: dict[str, Any] | None = None
    projectSnapshot: dict[str, Any] | None = None
    logs: list[TaskLogEntry] = field(default_factory=list)
    transcriptionProvider: str | None = None
    transcriptionModelSize: str | None = None
    transcriptionQualityPreset: str | None = None
    translationProvider: str | None = None
    translationModel: str | None = None
    outputMode: str | None = None

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "TaskHistoryRecord":
        task_mode = str(data.get("taskMode", "extractAndTranslate")).strip()
        if task_mode not in {"extractAndTranslate", "translateSubtitle"}:
            task_mode = "extractAndTranslate"

        engine_type = str(data.get("engineType", "subtitleImport")).strip()
        if engine_type not in {
            "cloudTranscription",
            "localTranscription",
            "subtitleImport",
        }:
            engine_type = "subtitleImport"

        status = str(data.get("status", "queued")).strip()
        if status not in {
            "queued",
            "transcribing",
            "translating",
            "editing",
            "exporting",
            "done",
            "error",
            "cancelled",
        }:
            status = "queued"

        return cls(
            taskId=str(data.get("taskId", "")).strip(),
            sourceFilePath=str(data.get("sourceFilePath", "")).strip(),
            sourceFileName=str(data.get("sourceFileName", "")).strip(),
            taskMode=task_mode,  # type: ignore[arg-type]
            sourceLanguage=str(data.get("sourceLanguage", "auto")).strip() or "auto",
            targetLanguage=str(data.get("targetLanguage", "zh-CN")).strip() or "zh-CN",
            outputFormats=[
                str(item).strip()
                for item in data.get("outputFormats", [])
                if str(item).strip()
            ],
            engineType=engine_type,  # type: ignore[arg-type]
            status=status,  # type: ignore[arg-type]
            createdAt=str(data.get("createdAt", "")).strip(),
            updatedAt=str(data.get("updatedAt", "")).strip(),
            exportPaths=[
                str(item).strip()
                for item in data.get("exportPaths", [])
                if str(item).strip()
            ],
            errorMessage=(
                str(data.get("errorMessage")).strip()
                if isinstance(data.get("errorMessage"), str)
                and str(data.get("errorMessage")).strip()
                else None
            ),
            subtitleSummary=SubtitleSummary.from_dict(data.get("subtitleSummary")),
            importSnapshot=data.get("importSnapshot")
            if isinstance(data.get("importSnapshot"), dict)
            else None,
            projectSnapshot=data.get("projectSnapshot")
            if isinstance(data.get("projectSnapshot"), dict)
            else None,
            logs=[
                TaskLogEntry.from_dict(item)
                for item in data.get("logs", [])
                if isinstance(item, dict)
            ],
            transcriptionProvider=(
                str(data.get("transcriptionProvider")).strip()
                if isinstance(data.get("transcriptionProvider"), str)
                and str(data.get("transcriptionProvider")).strip()
                else None
            ),
            transcriptionModelSize=(
                str(data.get("transcriptionModelSize")).strip()
                if isinstance(data.get("transcriptionModelSize"), str)
                and str(data.get("transcriptionModelSize")).strip()
                else None
            ),
            transcriptionQualityPreset=(
                str(data.get("transcriptionQualityPreset")).strip()
                if isinstance(data.get("transcriptionQualityPreset"), str)
                and str(data.get("transcriptionQualityPreset")).strip()
                else None
            ),
            translationProvider=(
                str(data.get("translationProvider")).strip()
                if isinstance(data.get("translationProvider"), str)
                and str(data.get("translationProvider")).strip()
                else None
            ),
            translationModel=(
                str(data.get("translationModel")).strip()
                if isinstance(data.get("translationModel"), str)
                and str(data.get("translationModel")).strip()
                else None
            ),
            outputMode=(
                str(data.get("outputMode")).strip()
                if isinstance(data.get("outputMode"), str)
                and str(data.get("outputMode")).strip()
                else None
            ),
        )


def _build_invalid_history_backup_path(history_path: Path) -> Path:
    timestamp = strftime("%Y%m%d-%H%M%S")
    return history_path.with_name(
        f"{history_path.stem}.invalid-{timestamp}{history_path.suffix}"
    )


def get_task_history_path() -> Path:
    config_path = get_config_path()
    return config_path.with_name("task-history.json")


def ensure_task_history_parent_exists() -> Path:
    history_path = get_task_history_path()
    history_path.parent.mkdir(parents=True, exist_ok=True)
    return history_path


def _reset_invalid_history(history_path: Path) -> list[TaskHistoryRecord]:
    if history_path.exists():
        backup_path = _build_invalid_history_backup_path(history_path)
        try:
            history_path.replace(backup_path)
        except OSError:
            try:
                backup_path.write_text(
                    history_path.read_text(encoding="utf-8", errors="replace"),
                    encoding="utf-8",
                )
                history_path.unlink(missing_ok=True)
            except OSError:
                pass

    save_task_history([])
    return []


def _sort_history(records: list[TaskHistoryRecord]) -> list[TaskHistoryRecord]:
    return sorted(records, key=lambda item: item.updatedAt or item.createdAt, reverse=True)


def load_task_history() -> list[TaskHistoryRecord]:
    history_path = ensure_task_history_parent_exists()
    if not history_path.exists():
        save_task_history([])
        return []

    try:
        with history_path.open("r", encoding="utf-8") as file:
            raw_data = json.load(file)
        if not isinstance(raw_data, list):
            raise ValueError("Task history JSON must be a list.")

        records = [
            TaskHistoryRecord.from_dict(item)
            for item in raw_data
            if isinstance(item, dict)
            and str(item.get("taskId", "")).strip()
            and str(item.get("sourceFilePath", "")).strip()
        ]
        return _sort_history(records)
    except (json.JSONDecodeError, OSError, TypeError, ValueError):
        return _reset_invalid_history(history_path)


def save_task_history(records: list[TaskHistoryRecord | dict[str, Any]]) -> list[TaskHistoryRecord]:
    normalized_records = [
        item if isinstance(item, TaskHistoryRecord) else TaskHistoryRecord.from_dict(item)
        for item in records
    ]
    history_path = ensure_task_history_parent_exists()

    trimmed_records = _sort_history(normalized_records)[:MAX_HISTORY_ITEMS]
    with history_path.open("w", encoding="utf-8") as file:
        json.dump(
            [item.to_dict() for item in trimmed_records],
            file,
            ensure_ascii=False,
            indent=2,
        )

    return trimmed_records


def upsert_task_history_record(record: TaskHistoryRecord | dict[str, Any]) -> TaskHistoryRecord:
    normalized_record = (
        record if isinstance(record, TaskHistoryRecord) else TaskHistoryRecord.from_dict(record)
    )
    if not normalized_record.taskId or not normalized_record.sourceFilePath:
        raise TaskHistoryError("Task history records must include taskId and sourceFilePath.")

    history = load_task_history()
    next_history: list[TaskHistoryRecord] = [normalized_record]
    next_history.extend(item for item in history if item.taskId != normalized_record.taskId)
    save_task_history(next_history)
    return normalized_record


loadTaskHistory = load_task_history
saveTaskHistory = save_task_history
upsertTaskHistoryRecord = upsert_task_history_record
